File: debug_specific_quality.ts
#!/usr/bin/env bun
/**
 * Debug specific quality scores for problematic measurements
 */

import { readFileSync } from "node:fs";
import { parse } from "csv-parse/sync";

// Import from typescript_lib
import {
  InMemoryStore,
  processMeasurement,
  type ProcessingResult,
} from "./typescript_lib/src/index";

// Import service layer
import {
  WeightProcessorService,
  type MeasurementInput,
} from "./services/weight_processor_service";

function getDefaultConfig(): any {
  return {
    database: {
      backend: "memory",
      table_name: "weight-processor-state",
      region: "us-east-1",
    },
    kalman: {
      initial_variance: 0.364,
      transition_covariance_weight: 0.018,
      transition_covariance_trend: 0.00012,
      observation_covariance: 5,
      reset: {
        initial: {
          enabled: true,
          initial_variance_multiplier: 10,
          weight_noise_multiplier: 50,
          trend_noise_multiplier: 50,
          observation_noise_multiplier: 20,
          adaptation_measurements: 10,
          adaptation_days: 10,
          adaptation_decay_rate: 2.5,
        },
        hard: {
          enabled: true,
          gap_threshold_days: 30,
          initial_variance_multiplier: 3,
          weight_noise_multiplier: 5,
          trend_noise_multiplier: 50,
          observation_noise_multiplier: 0.7,
          adaptation_measurements: 10,
          adaptation_days: 7,
          adaptation_decay_rate: 2.5,
        },
        soft: {
          enabled: true,
          min_weight_change_kg: 5,
          cooldown_days: 3,
          trigger_sources: ["questionnaire", "care-team-upload"],
          initial_variance_multiplier: 2,
          weight_noise_multiplier: 5,
          trend_noise_multiplier: 20,
          observation_noise_multiplier: 0.7,
          adaptation_measurements: 15,
          adaptation_days: 10,
          adaptation_decay_rate: 2.5,
        },
      },
    },
    quality_scoring: {
      use_harmonic_mean: true,
      threshold: 0.46,
      components: {
        kalman_fit: { weight: 0.4, enabled: true },
        temporal_consistency: { weight: 0.3, enabled: true },
        anomaly_detection: { weight: 0.2, enabled: true },
        source_reliability: { weight: 0.05, enabled: true },
        trend_alignment: { weight: 0.05, enabled: true },
      },
      component_weights: {
        kalman_fit: 0.4,
        temporal_consistency: 0.3,
        anomaly_detection: 0.2,
        source_reliability: 0.05,
        trend_alignment: 0.05,
      },
    },
    replay: {
      buffered_replay_enabled: true,
      buffer_hours: 24,
      max_buffer_measurements: 100,
    },
  };
}

async function main() {
  const csvFile = "/tmp/debug_user_full.csv";
  const targetIds = [
    "726b441f-eb43-47d9-8f3c-845d164e5a5b", // TS accepts, Py rejects
    "1a98b2c3-e023-4757-8d01-d35ef2fb363e", // Py accepts, TS rejects
  ];

  console.log("=== TypeScript Quality Score Debug ===");
  console.log(`Target IDs: ${targetIds.join(", ")}`);
  console.log("");

  // Load CSV
  const csvContent = readFileSync(csvFile, "utf-8");
  const rows = parse(csvContent, {
    columns: true,
    skip_empty_lines: true,
  });

  // Parse measurements
  const measurements: MeasurementInput[] = rows.map((row: any) => ({
    measurementId: row.id,
    weight: parseFloat(row.value_quantity),
    unit: row.unit || "kg",
    timestamp: new Date(row.effective_date_time || row.timestamp),
    source: row.source_type || "unknown",
  }));

  console.log(`Loaded ${measurements.length} measurements`);

  // Sort chronologically
  measurements.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());

  // Initialize storage and service
  const stateStore = new InMemoryStore();
  const config = getDefaultConfig();
  const service = new WeightProcessorService(stateStore, config);

  // Process batch
  const response = await service.processBatch("ADC64C0B-CB46-41F9-BDA0-CC11A35942D7", measurements);

  console.log(`\nProcessed ${response.measurements_processed} measurements`);
  console.log(`Accepted: ${response.measurements_accepted}, Rejected: ${response.measurements_rejected}`);
  console.log(`Replays triggered: ${response.replay_metadata?.length || 0}`);

  // Find our target measurements in results
  console.log("\n=== Target Measurement Results ===");
  for (const targetId of targetIds) {
    const measurementIndex = measurements.findIndex((m) => m.measurementId === targetId);
    if (measurementIndex >= 0) {
      const result = response.results[measurementIndex];
      const measurement = measurements[measurementIndex];
      console.log(`\nID: ${targetId}`);
      console.log(`  Timestamp: ${measurement.timestamp.toISOString()}`);
      console.log(`  Weight: ${measurement.weight} ${measurement.unit}`);
      console.log(`  Accepted: ${result.accepted}`);
      console.log(`  Quality Score: ${result.quality_score}`);
      console.log(`  Threshold: 0.46`);
      console.log(`  Score vs Threshold: ${result.quality_score - 0.46} (${result.quality_score >= 0.46 ? "PASS" : "FAIL"})`);

      if (result.quality_components) {
        console.log(`  Components:`);
        for (const [key, value] of Object.entries(result.quality_components)) {
          console.log(`    ${key}: ${value}`);
        }
      }

      if (!result.accepted) {
        console.log(`  Rejection reason: ${result.rejection_reason || result.reason || "Unknown"}`);
      }
    } else {
      console.log(`\n⚠️  ${targetId} not found in measurements`);
    }
  }
}

main().catch(console.error);

File: debug_specific_quality.py
#!/usr/bin/env python3
"""Debug specific quality scores for problematic measurements"""

import csv
from datetime import datetime, timezone
from pathlib import Path
import sys

# Add paths for imports
project_root = Path(__file__).parent
sys.path.insert(0, str(project_root / "python_lib" / "src"))
sys.path.insert(0, str(project_root / "be_implementation_service" / "src"))

# Import from be_implementation_service
from aws.api.models import Measurement
from aws.services.weight_processor_service import WeightProcessorService
from aws.config.config_manager import ConfigManager

# Import from python_lib
from weight_processor_lib.core.database import InMemoryStore

def parse_timestamp(date_str: str) -> datetime:
    """Parse timestamp string to datetime."""
    if not date_str:
        return datetime.now(timezone.utc)
    try:
        if "T" in date_str:
            dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
            if dt.tzinfo is None:
                return dt.replace(tzinfo=timezone.utc)
            return dt.astimezone(timezone.utc)
        elif " " in date_str:
            for fmt in ["%Y-%m-%d %H:%M:%S.%f", "%Y-%m-%d %H:%M:%S"]:
                try:
                    dt = datetime.strptime(date_str, fmt)
                    return dt.replace(tzinfo=timezone.utc)
                except ValueError:
                    continue
            raise ValueError(f"Cannot parse: {date_str}")
        else:
            dt = datetime.strptime(date_str, "%Y-%m-%d")
            return dt.replace(tzinfo=timezone.utc)
    except Exception:
        return datetime.now(timezone.utc)


def main():
    csv_file = "/tmp/debug_user_full.csv"
    target_ids = [
        "726b441f-eb43-47d9-8f3c-845d164e5a5b",  # TS accepts, Py rejects?
        "1a98b2c3-e023-4757-8d01-d35ef2fb363e",  # Py accepts, TS rejects?
    ]

    print("=== Python Quality Score Debug ===")
    print(f"Target IDs: {', '.join(target_ids)}")
    print("")

    # Load CSV
    measurements = []
    with open(csv_file, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            measurement = Measurement(
                uuid=row.get("id"),
                weight=float(row.get("value_quantity", "0")),
                unit=row.get("unit", "kg"),
                effectiveDateTime=parse_timestamp(row.get("effective_date_time") or row.get("timestamp")),
                source=row.get("source_type", "unknown"),
            )
            measurements.append(measurement)

    print(f"Loaded {len(measurements)} measurements")

    # Sort chronologically
    measurements = sorted(measurements, key=lambda m: m.measured_at)

    # Initialize storage and service
    state_store = InMemoryStore()
    config = ConfigManager.load_config(source="file")
    config["database"]["backend"] = "memory"
    service = WeightProcessorService(state_store=state_store, config=config)

    # Process batch
    response = service.process_batch("ADC64C0B-CB46-41F9-BDA0-CC11A35942D7", measurements)

    print(f"\nProcessed {response.measurements_processed} measurements")
    print(f"Accepted: {response.measurements_accepted}, Rejected: {response.measurements_rejected}")
    print(f"Replays triggered: {len(response.replay_metadata) if response.replay_metadata else 0}")

    # Find our target measurements in results
    print("\n=== Target Measurement Results ===")
    for target_id in target_ids:
        # Find measurement index
        measurement_index = None
        for i, m in enumerate(measurements):
            if m.measurement_id == target_id:
                measurement_index = i
                break

        if measurement_index is not None:
            result = response.results[measurement_index]
            measurement = measurements[measurement_index]
            print(f"\nID: {target_id}")
            print(f"  Timestamp: {measurement.measured_at.isoformat()}")
            print(f"  Weight: {measurement.weight_value} {measurement.weight_unit}")
            print(f"  Accepted: {result.accepted}")
            print(f"  Quality Score: {result.quality_score}")
            print(f"  Threshold: 0.46")

            if result.quality_score is not None:
                diff = result.quality_score - 0.46
                status = "PASS" if result.quality_score >= 0.46 else "FAIL"
                print(f"  Score vs Threshold: {diff:.15f} ({status})")
            else:
                print(f"  Score vs Threshold: N/A (quality_score is None)")

            if result.quality_components:
                print(f"  Components:")
                for key, value in result.quality_components.items():
                    print(f"    {key}: {value}")

            if not result.accepted:
                reason = result.rejection_reason or getattr(result, 'reason', 'Unknown')
                print(f"  Rejection reason: {reason}")
        else:
            print(f"\n⚠️  {target_id} not found in measurements")


if __name__ == "__main__":
    main()